crates/ide-assists/src/handlers/extract_function.rs

```diff
@@ -1264,6 +1276,28 @@ fn node_to_insert_after(body: &FunctionBody, anchor: Anchor) -> Option<SyntaxNod
     last_ancestor
 }
 
+fn find_non_trait_impl(trait_impl: &SyntaxNode) -> Option<SyntaxNode> {
+    let impl_type = Some(impl_type_name(trait_impl)?);
+
+    let mut sibblings = trait_impl.parent()?.children();
+    sibblings.find(|s| impl_type_name(s) == impl_type && !is_trait_impl(s))
+}
+
+fn last_impl_member(impl_node: &SyntaxNode) -> Option<SyntaxNode> {
+    impl_node.children().find(|c| c.kind() == SyntaxKind::ASSOC_ITEM_LIST)?.last_child()
+}
+
+fn is_trait_impl(node: &SyntaxNode) -> bool {
+    match ast::Impl::cast(node.clone()) {
```

Comment:
I was hoping to avoid the clones here and on [l1298](https://github.com/rust-lang/rust-analyzer/pull/12991/files#diff-8542dad26a826a4d7012b39913b333fbc3c505cf736d3158a855a7065aa0dffaR1298), but the issue is that `ast::Impl::cast` requires ownership of the node. I could implement these methods directly on `SyntaxNode`, without using `ast::Impl`. This would avoid cloning, but also makes the resulting code less clear and more fragile.

Reply:
As a quick and dirty optimization, I think you can check the `kind()` first.

Reply:
I'll add that. Just hoping there was a way to completely avoid the clone. It feels a bit silly, given that we're only using it read-only and even then are only interested in a small part of the node.

Reply:
Note that cloning nodes is very cheap as it's just a ref count bump